libs/gl-client/src/pairing/service.rs

```diff
@@ -0,0 +1,345 @@
+        Ok(self
+            .client
+            .clone()
+            .get_pairing_data(GetPairingDataRequest {
+                session_id: session_id.to_string(),
+            })
+            .await?
+            .into_inner())
+    }
+
+    pub async fn approve_pairing(
+        &self,
+        session_id: &str,
+        node_id: &[u8],
+        device_name: &str,
+        restrs: &str,
+    ) -> Result<Empty, PairingError> {
+        if let Some(rune) = self.rune.clone() {
+            if let Some(key) = &self.key {
+                let timestamp = SystemTime::now()
+                    .duration_since(UNIX_EPOCH)
+                    .map_err(|e| PairingError::ApproveParingError(e.to_string()))?
+                    .as_secs();
+
+                // Gather data to sign over.
+                let mut buf = vec![];
+                buf.put(session_id.as_bytes());
+                buf.put_u64(timestamp);
+                buf.put(&node_id[..]);
+                buf.put(device_name.as_bytes());
+                buf.put(restrs.as_bytes());
+
+                // Sign data.
+                let key = {
+                    let mut key = std::io::Cursor::new(key);
+                    pemfile::pkcs8_private_keys(&mut key)
+                        .map_err(|e| PairingError::ApproveParingError(e.to_string()))?
+                        .remove(0)
+                };
+                let kp = EcdsaKeyPair::from_pkcs8(
+                    &signature::ECDSA_P256_SHA256_FIXED_SIGNING,
+                    key.as_ref(),
+                )
+                .map_err(|e| {
+                    PairingError::ApproveParingError(format!(
+                        "could not get keypair from PEM string: {}",
+                        e
+                    ))
+                })?;
+                let rng = rand::SystemRandom::new();
+                let sig = kp
+                    .sign(&rng, &buf)
+                    .map_err(|e| PairingError::ApproveParingError(e.to_string()))?
+                    .as_ref()
+                    .to_vec();
+
+                // Send approval.
+                Ok(self
+                    .client
+                    .clone()
+                    .approve_pairing(ApprovePairingRequest {
+                        session_id: session_id.to_string(),
+                        timestamp,
+                        node_id: node_id.to_vec(),
+                        device_name: device_name.to_string(),
+                        restrs: restrs.to_string(),
+                        sig: sig,
+                        rune: rune.to_base64(),
+                    })
+                    .await?
+                    .into_inner())
+            } else {
+                Err(PairingError::ApproveParingError(
+                    "tls key missing".to_string(),
+                ))
+            }
+        } else {
+            Err(PairingError::ApproveParingError("rune missing".to_string()))
+        }
```

Comment:
Move this to the top:

```
if self.rune.is_none() {
    return Err(...)
}
// Continue with one fewer indendation.
```

The reason is that in long indented blocks are hard to see the matching indentations, and having to care about an early check at the end, when really it could just have been an early exit is just strange :-)